File: samples/node-user-management-svc/src/service/UserService.js
import { body, validationResult } from 'express-validator';
import { UserRepository } from '../repo/UserRepository.js';
import { PasswordSecurityService } from '../security/PasswordSecurityService.js';
import { EventBus } from '../events/EventBus.js';
import { ConflictException, BadRequestException } from '../error/handlers.js';

// Validation chain reflecting spec rules
export const createUserValidators = [
  body('email').isEmail().withMessage('email must be a valid email'),
  body('password').isString().isLength({ min: 8 }).withMessage('password min length 8'),
  body('displayName').isString().isLength({ min: 1, max: 50 }).withMessage('displayName required, max 50'),
];

export async function createUser(userDTO) {
  // Additional runtime guard (controllers already run validators)
  if (!userDTO || typeof userDTO !== 'object') throw new BadRequestException('Invalid body');

  // decision: existsByEmail
  const exists = UserRepository.existsByEmail(userDTO.email);
  if (exists) {
    throw new ConflictException('A user with this email already exists.');
  }

  // domain-service-call: hashPassword
  const hashedPassword = PasswordSecurityService.hashPassword(userDTO.password);

  // mapping: DTO -> UserEntity
  const nowIso = new Date().toISOString();
  const newUserEntity = Object.freeze({
    email: String(userDTO.email),
    passwordHash: hashedPassword,
    displayName: String(userDTO.displayName),
    createdAt: nowIso,
  });

  // persistence-call: save
  const savedUser = UserRepository.save(newUserEntity);

  // mapping: event payload
  const eventPayload = Object.freeze({ userId: savedUser.id, email: savedUser.email });

  // publish-event: UserCreatedEvent
  EventBus.publish('UserCreatedEvent', eventPayload);

  // mapping: entity -> response DTO
  const responseDTO = Object.freeze({
    id: savedUser.id,
    email: savedUser.email,
    displayName: savedUser.displayName,
    createdAt: savedUser.createdAt,
  });

  // return
  return responseDTO;
}

export const UserService = Object.freeze({ createUser, createUserValidators });

File: samples/user-management-svc/src/main/java/com/example/usermanagement/service/UserService.java
package com.example.usermanagement.service;

import com.example.usermanagement.domain.UserEntity;
import com.example.usermanagement.dto.UserCreateDTO;
import com.example.usermanagement.dto.UserResponseDTO;
import com.example.usermanagement.events.UserCreatedEvent;
import com.example.usermanagement.repo.UserRepository;
import com.example.usermanagement.security.PasswordSecurityService;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.OffsetDateTime;
import java.util.UUID;

@Service
public class UserService {
    private final UserRepository userRepository;
    private final PasswordSecurityService passwordSecurityService;
    private final ApplicationEventPublisher eventPublisher;

    public UserService(UserRepository userRepository,
                       PasswordSecurityService passwordSecurityService,
                       ApplicationEventPublisher eventPublisher) {
        this.userRepository = userRepository;
        this.passwordSecurityService = passwordSecurityService;
        this.eventPublisher = eventPublisher;
    }

    @Transactional
    public UserResponseDTO createUser(UserCreateDTO userDTO) {
        // validation is additionally enforced by annotations; basic checks
        if (userRepository.existsByEmail(userDTO.getEmail())) {
            throw new ConflictException("A user with this email already exists.");
        }

        String hashedPassword = passwordSecurityService.hashPassword(userDTO.getPassword());

        UserEntity newUser = UserEntity.builder()
                .id(UUID.randomUUID())
                .email(userDTO.getEmail())
                .passwordHash(hashedPassword)
                .displayName(userDTO.getDisplayName())
                // createdAt handled by @CreationTimestamp, but set now() as per spec mapping
                .createdAt(OffsetDateTime.now())
                .build();

        UserEntity saved = userRepository.save(newUser);

        // publish event
        eventPublisher.publishEvent(new UserCreatedEvent(this, saved.getId(), saved.getEmail()));

        return UserResponseDTO.builder()
                .id(saved.getId())
                .email(saved.getEmail())
                .displayName(saved.getDisplayName())
                .createdAt(saved.getCreatedAt())
                .build();
    }

    public static class ConflictException extends RuntimeException {
        public ConflictException(String message) { super(message); }
    }
}
